src/Utilities/Compiler/Options/SymbolNamesWithValueOption.cs

```diff
@@ -43,89 +81,287 @@ private SymbolNamesWithValueOption()
                         }
+                    }
 
-                        if (!symbolsBuilder.ContainsKey(symbol))
-                        {
-                            symbolsBuilder.Add(symbol, parts.AssociatedValue);
-                        }
+                    if (!symbolsBuilder.ContainsKey(symbol))
+                    {
+                        symbolsBuilder.Add(symbol, parts.AssociatedValue);
                     }
                 }
             }
-
-            if (namesBuilder.Count == 0 && symbolsBuilder.Count == 0)
-            {
-                return Empty;
-            }
-
-            return new SymbolNamesWithValueOption<TValue>(namesBuilder.ToImmutableDictionaryAndFree(), symbolsBuilder.ToImmutableDictionaryAndFree());
         }
 
         public bool IsEmpty => ReferenceEquals(this, Empty);
 
         public bool Contains(ISymbol symbol)
-            => _symbols.ContainsKey(symbol) || _names.ContainsKey(symbol.Name);
+            => _symbols.ContainsKey(symbol) || _names.ContainsKey(symbol.Name) || TryGetFirstWildcardMatch(symbol, out _);
 
         /// <summary>
         /// Gets the value associated with the specified symbol in the option specification.
         /// </summary>
-        public bool TryGetValue(ISymbol symbol, [NotNullWhen(true)] out TValue value) =>
-            _symbols.TryGetValue(symbol, out value) || _names.TryGetValue(symbol.Name, out value);
+        public bool TryGetValue(ISymbol symbol, [NotNullWhen(true)] out TValue value)
+        {
+            if (_symbols.TryGetValue(symbol, out value) || _names.TryGetValue(symbol.Name, out value))
+            {
+                return true;
+            }
+
+            if (TryGetFirstWildcardMatch(symbol, out var match))
+            {
+                value = match.Value;
+                return true;
+            }
+
+#pragma warning disable CS8653 // A default expression introduces a null value for a type parameter.
+            value = default;
+#pragma warning restore CS8653 // A default expression introduces a null value for a type parameter.
+            return false;
+        }
 
         public override bool Equals(object obj) => Equals(obj as SymbolNamesWithValueOption<TValue>);
 
         public bool Equals(SymbolNamesWithValueOption<TValue>? other)
-            => other != null && _names.IsEqualTo(other._names) && _symbols.IsEqualTo(other._symbols);
+            => other != null && _names.IsEqualTo(other._names) && _symbols.IsEqualTo(other._symbols) && _wildcardNamesBySymbolKind.IsEqualTo(other._wildcardNamesBySymbolKind);
 
         public override int GetHashCode()
-            => HashUtilities.Combine(HashUtilities.Combine(_names), HashUtilities.Combine(_symbols));
+            => HashUtilities.Combine(HashUtilities.Combine(_names), HashUtilities.Combine(_symbols), HashUtilities.Combine(_wildcardNamesBySymbolKind));
+
+        private bool TryGetFirstWildcardMatch(ISymbol symbol, out KeyValuePair<string, TValue> firstMatch)
```

Comment:
I see a bunch of return statements in this method where nothing is added/updated in the cache, while some of them do. It is also not clear if the intention is to cache values only for namespace and type symbols or for all symbols. I think former might be more suitable to keep the cache size to reasonable.

Reply:
You are correct I have missed some parts and I forgot to reuse the cache at the various steps.

Reply:
I still find the logic very confusing. We should not have so many points in code where you are updating or reading from the cache. There should be a single GetOrCompute point at the top which accesses the cache, and then invokes a core compute function which does not access/update the cache. Let me know if you'd prefer if I create a simple commit on top of your branch to convey what I mean.

Reply:
I haven't finished the refactoring, for example methods and indexers properties are matching is still not working properly so I will do one more commit to try and handle this. 

I am more than happy for you to do (a) commit(s) on this PR or any. This is exactly the kind of "problem solving" situation where a chat and a whiteboard are useful.

Reply:
> for example methods and indexers properties are matching is still not working properly

I think the hot path here is for containing types and symbols. IMO, it should be fine to start with caching just those symbol kinds.

Reply:
@Evangelink I think I have blocked this PR for long enough now without much action from my side. Lets do the following to get traction here:

1. Given this code path is only taken when user provides a wild card option value, it will not affect normal analysis scenarios. So any performance concern here is non-blocking for the user. I think that is sufficient to postpone the performance tuning and cleanup work for this code path.
2. Can you please file a tracking issue with following notes and tag me?
   1. Consider revisiting this code for code refactoring/cleanup for easier future maintenance.
   2. Consider tuning the cache size and hit/misses here to improve performance.
3. Please resolve the merge conflicts on the PR. I am going to mark this PR approved and move forward with the merge.

Reply:
See #3403